raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -983,48 +999,88 @@ private FetchResponseData tryCompleteFetchRequest(
 
-            if (state.updateReplicaState(replicaId, currentTimeMs, info.startOffsetMetadata)) {
-                onUpdateLeaderHighWatermark(state, currentTimeMs);
+                if (state.updateReplicaState(replicaId, currentTimeMs, info.startOffsetMetadata)) {
+                    onUpdateLeaderHighWatermark(state, currentTimeMs);
+                }
+
+                records = info.records;
+            } else {
+                records = MemoryRecords.EMPTY;
             }
 
-            return buildFetchResponse(Errors.NONE, info.records, Optional.empty(), state.highWatermark());
+            return buildFetchResponse(Errors.NONE, records, validatedOffsetAndEpoch, state.highWatermark());
+        } catch (Exception e) {
+            logger.error("Caught unexpected error in fetch completion of request {}", request, e);
+            return buildEmptyFetchResponse(Errors.UNKNOWN_SERVER_ERROR, Optional.empty());
         }
     }
 
     /**
      * Check whether a fetch offset and epoch is valid. Return the diverging epoch, which
      * is the largest epoch such that subsequent records are known to diverge.
      */
-    private Optional<OffsetAndEpoch> validateFetchOffsetAndEpoch(long fetchOffset, int lastFetchedEpoch) {
-        if (fetchOffset == 0 && lastFetchedEpoch == 0) {
-            return Optional.empty();
+    private ValidatedFetchOffsetAndEpoch validateFetchOffsetAndEpoch(long fetchOffset, int lastFetchedEpoch) {
+        if (log.startOffset() == 0 && fetchOffset == 0) {
+            if (lastFetchedEpoch != 0) {
+                logger.warn(
+                    "Replica sent a zero fetch offset ({}) but the last fetched epoch ({}) was not zero",
+                    fetchOffset,
+                    lastFetchedEpoch
+                );
+            }
+            return ValidatedFetchOffsetAndEpoch.valid(new OffsetAndEpoch(fetchOffset, lastFetchedEpoch));
         }
 
-        OffsetAndEpoch endOffsetAndEpoch = log.endOffsetForEpoch(lastFetchedEpoch)
-            .orElse(new OffsetAndEpoch(-1L, -1));
-        if (endOffsetAndEpoch.epoch != lastFetchedEpoch || endOffsetAndEpoch.offset < fetchOffset) {
-            return Optional.of(endOffsetAndEpoch);
+
+        OffsetAndEpoch endOffsetAndEpoch = log.endOffsetForEpoch(lastFetchedEpoch).orElseThrow(() -> {
+            return new IllegalStateException(
+                String.format(
+                    "Expected to find an end offset for epoch %s since it must be less than the current epoch %s",
+                    lastFetchedEpoch,
+                    quorum.epoch()
+                )
+            );
+        });
+
+        if (log.oldestSnapshotId().isPresent() &&
+            ((fetchOffset < log.startOffset()) ||
+             (fetchOffset == log.startOffset() && lastFetchedEpoch != log.oldestSnapshotId().get().epoch) ||
+             (lastFetchedEpoch < log.oldestSnapshotId().get().epoch))) {
```

Comment:
Related to my other comment, but I think this check can go away if we change `endOffsetForEpoch` to only return the end offset and epoch when they can be definitively determined.

The other condition where we need to return a snapshot is when the returned epoch `endOffsetAndEpoch.epoch` equals the oldest snapshot epoch and the fetch offset is less than the oldest snapshot offset, which I think is equivalent to what is here. So I guess I'm thinking we can write this like this: 
```

OffsetAndEpoch endOffsetAndEpoch = log.endOffsetForEpoch(lastFetchedEpoch)
if (!endOffsetAndEpoch.isPresent) {
  // return latest snapshot
}

OffsetAndEpoch oldestSnapshotId = log.oldestSnapshotId().getOrElse(new OffsetAndEpoch(0, 0));
if (endOffsetAndEpoch.epoch == oldestSnapshotId.epoch && fetchOffset < oldestSnapshotId.offset) {
  // return latest snapshot
else if (endOffsetAndEpoch.epoch != lastFetchedEpoch || endOffsetAndEpoch.offset < fetchOffset) {
  // return diverging epoch
} else {
  // valid fetch
}
```

Reply:
Hmmm. The case that I have in mind is this one:

```
oldestSnapshotId: OffsetAndEpoch(10, 2)
leaderEpochCache: (epoch: 4, startOffset: 10), (epoch: 5, starOffset: 20), ...
```

If the follower fetches `(fetchOffset: ..., lastFetchdEpoch: 3)`, then the leader should return `diverging(10, 2)`.

If the follower fetches `(fetchOffset: 15, lastFetchdEpoch: 2)`, then the leader should return `diverging(10, 2)`.

I think we can only do this if we have the check against the `oldestSnapshotId`.